File: questions/power_set/solution/power_set.py

def power_set(values):
    """Returns all the power sets of a list

        >>> print power_set(['a','b','c'])
        [[], ['c'], ['b'], ['b', 'c'], ['a'], ['a', 'c'], ['a', 'b'], ['a', 'b', 'c']]

        >>> power_set([1,2])
        [[], [2], [1], [1, 2]]

        >>> power_set([])
        [[]]
    """

    if not values:
        return [[]]

    remaining = power_set(values[1:])
    return remaining + [[values[0]] + x for x in remaining]


print power_set(['a', 'b', 'c'])


#####################################################################
# Don't touch the code below!
# This allows the doctests to run
if __name__ == '__main__':
    import doctest
    if doctest.testmod().failed == 0:
        print "\n*** ALL TESTS PASSED. AWESOME WORK!\n"

File: questions/power_set/solution/power_set.cpp
#include <vector>
#include <iostream>

using namespace std;

template<typename T>
vector<vector<T>> powSet(vector<T> set, int n){    
    if(n == set.size()){
        return vector<vector<T>>{ vector<T>() };
    }
    vector<vector<T>> all;
    vector<vector<T>> sets = powSet(set, n+1);
    int val = set[n];
    for(auto s: sets){
        vector<T> cp(s);
        cp.push_back(val);
        all.push_back(s);
        all.push_back(cp);
    }
    return all;
}

template<typename T>
void print(const vector<vector<T>>& powset){
    cout << "power set: " << endl;
    for(vector<T> s : powset){
        cout << "{";
        for(T i : s)cout << i << ",";
        cout << "}, ";
    }
    cout << endl;
}

int main(){
    auto ps = powSet<int>(vector<int>{1, 2, 3, 4}, 0 );
    print<int>(ps);

    auto chps = powSet<char>(vector<char>{'a', 'b', 'c', 'd'}, 0 );
    print<char>(chps);
    return 0;
}
